File: year_2021/solutions/julia/day_06.jl
include(joinpath(@__DIR__(), "utils", "files.jl"))


function count_fish_after_n_days(initial_fish_population::Array{Int}, n::Int)::Int
    population_by_timer = Dict{Int, Int}([(timer, count(==(timer), initial_fish_population)) for timer in 0:8])
    for _ in 1:n
        new_population = Dict{Int, Int}()
        for timer in 0:8
            if timer == 8
                new_population[timer] = population_by_timer[0]
            elseif timer == 6
                new_population[timer] = population_by_timer[0] + population_by_timer[7]
            else
                new_population[timer] = population_by_timer[timer + 1]
            end
        end
        population_by_timer = new_population
    end
    return sum(values(population_by_timer))
end


function main()
    input_file_path::String = joinpath(INPUTS_FOLDER, splitext(basename(@__FILE__))[1], "input.txt")
    fish_population::Array{Int} = parse.(Int, split(read(open(input_file_path), String), ","))

    # Part 1
    part_1_result::Int = count_fish_after_n_days(fish_population, 80)
    @assert part_1_result == 391671
    println("Part 1 result : ", part_1_result)

    # Part 2
    part_2_result::Int = count_fish_after_n_days(fish_population, 256)
    @assert part_2_result == 1754000560399
    println("Part 2 result : ", part_2_result)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: year_2021/solutions/python/day_06.py
from pathlib import Path
from typing import List

from year_2021.solutions.python.utils.files import INPUTS_FOLDER, read_int_line


def count_fish_after_n_days(fish_population: List[int], n: int) -> int:
    population_by_timer = {timer: fish_population.count(timer) for timer in range(0, 9)}
    for _ in range(n):
        new_population = {}
        for timer in range(0, 9):
            match timer:
                case 8:
                    new_population[timer] = population_by_timer[0]
                case 6:
                    new_population[timer] = population_by_timer[0] + population_by_timer[7]
                case _:
                    new_population[timer] = population_by_timer[timer + 1]
        population_by_timer = new_population
    return sum(population_by_timer.values())


if __name__ == '__main__':
    input_file_path = INPUTS_FOLDER / Path(__file__).with_suffix('').name / 'input.txt'
    initial_fish_population: List[int] = read_int_line(input_file_path=input_file_path)

    # Part 1
    part_1_result: int = count_fish_after_n_days(fish_population=initial_fish_population, n=80)
    assert part_1_result == 391671
    print('Part 1 result :', part_1_result)

    # Part 2
    part_2_result: int = count_fish_after_n_days(fish_population=initial_fish_population, n=256)
    assert part_2_result == 1754000560399
    print('Part 2 result :', part_2_result)
